File: medium/6. Longest Peak/longest_peak.cpp
#include <vector>

int longestPeak(std::vector<int> array) {
  int longestPeakLength = 0;
  int i = 1;
  while (i < int(array.size() - 1)) {
    if (!(array[i - 1] < array[i] && array[i + 1] < array[i])) {
      i++;
      continue;
    }

    int leftIdx = i - 2;
    while (leftIdx >= 0 && array[leftIdx] < array[leftIdx + 1]) {
      leftIdx--;
    }

    int rightIdx = i + 2;
    while (rightIdx < int(array.size()) &&
           array[rightIdx] < array[rightIdx - 1]) {
      rightIdx++;
    }

    int currentPeakLength = rightIdx - leftIdx - 1;
    longestPeakLength = std::max(longestPeakLength, currentPeakLength);
    i = rightIdx;
  }
  return longestPeakLength;
}

File: dsa/medium/6. Longest Peak/longest_peak.go
package main

func LongestPeak(array []int) int {
	longestPeakLength := 0
	i := 1

	for i < len(array)-1 {
		if !(array[i-1] < array[i] && array[i+1] < array[i]) {
			i++
			continue
		}
		left := i - 2
		for left >= 0 && array[left] < array[left+1] {
			left--
		}

		right := i + 2
		for right < len(array) && array[right] < array[right-1] {
			right++
		}
		currentPeakLength := right - left - 1
		if currentPeakLength > longestPeakLength {
			longestPeakLength = currentPeakLength
		}
		i = right
	}
	return longestPeakLength
}
